fabric-api-lookup-api-v1/src/main/java/net/fabricmc/fabric/impl/lookup/item/ItemKeyImpl.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.lookup.item;
+
+import java.util.Objects;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.item.Item;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+
+import net.fabricmc.fabric.api.lookup.v1.item.ItemKey;
+
+public class ItemKeyImpl implements ItemKey {
+	private final Item item;
+	private final @Nullable CompoundTag tag;
+	private final int hashCode;
+
+	ItemKeyImpl(Item item, CompoundTag tag) {
+		this.item = item;
+		this.tag = tag == null ? null : tag.copy(); // defensive copy
+		hashCode = Objects.hash(item, tag);
+	}
+
+	@Override
+	public Item getItem() {
+		return item;
+	}
+
+	@Override
+	public boolean hasTag() {
+		return tag != null;
+	}
+
+	@Override
+	public boolean tagMatches(@Nullable CompoundTag other) {
+		return Objects.equals(tag, other);
+	}
+
+	@Override
+	public @Nullable CompoundTag copyTag() {
+		return tag == null ? null : tag.copy();
+	}
+
+	@Override
+	public CompoundTag toTag() {
+		CompoundTag result = new CompoundTag();
+		result.putString("item", Registry.ITEM.getId(item).toString());
+
+		if (tag != null) {
+			result.put("tag", tag.copy());
+		}
```

Comment:
why not 
```suggestion		
		result.put("tag", copyTag());		
```
?

Reply:
That's better considering a stack with no tag and a stack with null tag are different.

Reply:
Why no that's precisely why I didn't do it.

Reply:
We don't need to waste storage space on an empty tag.

Reply:
Yeah I just realized it